memory/src/allocator/dynamic.rs

```
@@ -611,20 +630,33 @@ where
         self.blocks == 0
     }
 
-    fn acquire_blocks(&mut self, count: u32) -> Option<u32> {
+    fn acquire_blocks(&mut self, count: u32, block_size: u64, align: u64) -> Option<u32> {
         debug_assert!(count > 0 && count <= MAX_BLOCKS_PER_CHUNK);
+
+        // Find out if we have `count` consecutive free blocks.
         let mut blocks = !0;
         for i in 0..count {
             blocks &= self.blocks >> i;
         }
+        // Index of where the free blocks start.
         let index = blocks.trailing_zeros();
-        if index == MAX_BLOCKS_PER_CHUNK {
-            None
-        } else {
-            let mask = ((1 << count) - 1) << index;
-            self.blocks &= !mask;
-            Some(index)
+
+        // We may not be aligned at `index`, but may find alignment
+        // at subsequent indices.
+        for i in index..MAX_BLOCKS_PER_CHUNK {
+            let mask: u64 = ((1 << count) - 1) << i;
+
+            // Exit if we don't have enough free blocks at the current offset.
+            if (self.blocks & mask).count_ones() < count {
+                return None;
```

If check fails we should look further instead of returning.
For example if `self.blocks` is `b000110110` and we need `2` blocks with `align = 2 * block_size` then first pair doesn't fit, but this check will fail at next indicex (and at one after it'd fail too) and another free pair that statisfies requirements won't be examined.